### PR TITLE
Do not define opaque types when selecting impls

### LINKED ISSUE #126117: Regression in nightly-2024-05-27
The code in my hobby project Prism compiles on stable, but not on the latest nightly.
https://github.com/JonathanBrouwer/Prism/tree/222637a80cc2bb7794d0e4f649665fdd4ca23320
Sadly the codebase is quite large and I don't really see a good way to reduce it.

I ran cargo-bisect-rustc, which got the following output: (Which is a version bump commit, which confuses me)

searched nightlies: from nightly-2024-01-01 to nightly-2024-06-07
regressed nightly: nightly-2024-03-18
searched commit range: https://github.com/rust-lang/rust/compare/766bdce744d531267d53ba2a3f9ffcda69fb9b17...eb45c844407968ea54df0d9870ebce9e3235b706
regressed commit: https://github.com/rust-lang/rust/commit/4c1b9c3e2f402efe27a24e424c9b59aa2fa5c83f

The most interesting error that is returned on nightly is:
```
error[E0391]: cycle detected when computing type of opaque `parser::parser_rule_body::parser_body_sub_annotations::{opaque#0}`
   --> prism-parser/src/parser/parser_rule_body.rs:122:6
    |
122 | ) -> impl Parser<'arn, 'grm, Cow<'arn, ActionResult<'arn, 'grm>>, E> + 'a {
    |      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
note: ...which requires type-checking `parser::parser_rule_body::parser_body_sub_annotations`...
   --> prism-parser/src/parser/parser_rule_body.rs:125:27
    |
125 |             let mut res = parser_body_sub_annotations(rules, blocks, rest, expr, rule_args, vars)
    |                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    = note: ...which requires evaluating trait selection obligation `parser::parser_rule_body::parser_body_sub_annotations::{opaque#0}: core::marker::Unpin`...
    = note: ...which again requires computing type of opaque `parser::parser_rule_body::parser_body_sub_annotations::{opaque#0}`, completing the cycle
note: cycle used when computing type of `parser::parser_rule_body::parser_body_sub_annotations::{opaque#0}`
   --> prism-parser/src/parser/parser_rule_body.rs:122:6
    |
122 | ) -> impl Parser<'arn, 'grm, Cow<'arn, ActionResult<'arn, 'grm>>, E> + 'a {
    |      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    = note: see https://rustc-dev-guide.rust-lang.org/overview.html#queries and https://rustc-dev-guide.rust-lang.org/query.html for more information
```
Though multiple errors are returned. The entire error log is:
https://gist.github.com/JonathanBrouwer/33d429eab435c266029b3a0828642760

### PR DESCRIPTION
**Pull request by oli-obk**
fixes #126117

r? @lcnr for inconsistency with next solver

### CODE REVIEW

**Comment by oli-obk** [comment]
probably needs to be backported to beta (I think we cut this weekend?)

File: tests/ui/type-alias-impl-trait/issue-84660-unsoundness.rs [hunk 1]
```diff
@@ -1,10 +1,6 @@
 // Another example from issue #84660, this time weaponized as a safe transmute: an opaque type in an
 // impl header being accepted was used to create unsoundness.
 
-//@ revisions: current next
-//@ ignore-compare-mode-next-solver (explicit revisions)
-//@[next] compile-flags: -Znext-solver
```
Comment (lcnr) [inline tests/ui/type-alias-impl-trait/issue-84660-unsoundness.rs:6]: please keep these revisions

File: tests/ui/type-alias-impl-trait/nested-tait-inference.rs [hunk 2]
```diff
@@ -1,21 +1,18 @@
 #![feature(type_alias_impl_trait)]
 #![allow(dead_code)]
 
-//@ revisions: current next
-//@ ignore-compare-mode-next-solver (explicit revisions)
-//@[next] compile-flags: -Znext-solver
-//@check-pass
```
Comment (lcnr) [inline tests/ui/type-alias-impl-trait/nested-tait-inference.rs:7]: same

File: tests/ui/type-alias-impl-trait/nested-tait-inference2.rs [hunk 3]
```diff
@@ -1,10 +1,6 @@
 #![feature(type_alias_impl_trait)]
 #![allow(dead_code)]
 
-//@ revisions: current next
-//@ ignore-compare-mode-next-solver (explicit revisions)
-//@[next] compile-flags: -Znext-solver
```
Comment (lcnr) [inline tests/ui/type-alias-impl-trait/nested-tait-inference2.rs:6]: same

**Review by lcnr — COMMENTED** [review]
r=me after undoing the revision changes

**Comment by apiraino** [comment]
Beta backport approved as per compiler team [on Zulip](https://rust-lang.zulipchat.com/#narrow/stream/238009-t-compiler.2Fmeetings/topic/.5Bweekly.5D.202024-06-13/near/444476467). A backport PR will be authored by the release team at the end of the current development cycle.

@rustbot label +beta-accepted

**Pull request merged**